WT-5911 Remove unsafe calls to verify HS from session->verify.

sueloverso opened the pull request.

sueloverso commented:
@agorrod you should review this because it removes/moves some functionality. I removed the `history_store` configuration setting for `verify`. Calling that function was not safe as the table was subject to change out from underneath the code. Verifying the overall history store is now part of wiredtiger_open when salvage is configured. That part will likely change in some way in WT-5841 (which is next on my ticket list).

keithbostic commented:
@sueloverso, I've added @sulabhM as a reviewer, since he (at least), was involved in previous decisions about how and when to verify the history store file.

@sulabhM, @agorrod, once you two sign off on this change, I'm happy to do the detailed review, if you don't want to.

sueloverso commented:
With this change the CONFIG in the ticket, which was failing 100% of the time, has now passed several iterations in a row.

sueloverso commented:
@agorrod and @sulabhM while I have relocked this PR because it needs more work you should take a first look at it because it removes some functionality that I think is necessary. The form of `wt_verify_history_store` is also likely to change and be split. However if you don't agree with the general path we should discuss sooner rather than later.

sulabhM commented:
I have left a comment on WT-5911, saying that I like this idea that @sueloverso is headed towards. Either Vamsi has already put some effort in an overall verify discussion or was planning to, so it might be useful for all of us to get in sync with the approach to follow before putting a lot of effort in the code change.

agorrod commented:
@sueloverso I added a comment to WT-5911, but I haven't looked at the code change here. Should I take a closer look?

sueloverso commented:
@agorrod @sulabhM and @keithbostic although it still has bugs (verifying a btree as part of `wt_verify` is leaving a hazard pointer set), this is the expected form of this change. It would be a good time for a high level pass. The new `verify_metadata` config to `wiredtiger_open` is entirely untested at the moment.

agorrod commented on dist/api_data.py:
```diff
@@ -1034,6 +1034,10 @@ def __ge__(self, other):
         ]),
+    Config('verify_metadata', 'false', r'''
+        open connection and verify any WiredTiger-owned table. This API
+        allows verification and detection of corruption in WiredTiger tables.''',
```
Does calling `wiredtiger_open` with this flag set result in a valid open connection once verify has completed?

I know why we got here, but I still think a method on the `WT_CONNECTION` is a more natural API for users. How much harder is that to implement?

keithbostic replied:
I'll defer to @sueloverso, but I think it's going to be quite difficult.

We'd have to somehow split apart the existing `wiredtiger_open()` handling of these files, as it opens and reads them, and we'd also have to guarantee no other methods were called between the `wiredtiger_open()` call and the new verify method call, which implies some kind of locking.

What are the semantics if `wiredtiger_open()` succeeds in opening these files, but then they fail verify in the new method?

I could be totally missing an approach (??), but it seems hard to me.

sueloverso replied:
I agree with @keithbostic. We have to verify before opening the history store or metadata file. We cannot verify them after they're open. That is the whole problem with what exists today. That is why it is strategically called before the metadata is opened the first time. We don't really have a useful connection if the metadata and history store don't verify.

agorrod commented on src/utilities/util_verify.c:
```diff
@@ -17,17 +17,12 @@ util_verify(WT_SESSION *session, int argc, char *argv[])
     size_t size;
     int ch;
     char *config, *dump_offsets, *uri;
-    bool dump_address, dump_blocks, dump_layout, dump_pages, dump_history, hs_verify,
```
I think there should be an obvious way to verify the metadata via the command line utility. It's probably possible to do that via -C with your changes, but I'd prefer something more explicit in here.

sueloverso replied:
I don't object to this. It is slightly unusual in that a verify of the metadata will effectively be a `wiredtiger_open` followed immediately by `connection->close`. That is okay. With the -C option, the user could do some other command.

sueloverso commented:
@keithbostic please consider the changes to the `wt` command. For @agorrod I added the `-m` option to verify the metadata. But I also changed it so that can be the only thing a user chooses to do. That is, that allows the user to use it like `wt -h DIR -m` without any additional command (or with a command if desired). It would have required a lot of different manipulation to add `wt verify -m` because by the time we get into `util_verify` we're long past `wiredtiger_open` and parsing the options to an individual command. So it would make verify a specially processed command before `wiredtiger_open` to go that way.

keithbostic commented on src/history/hs.c:
```diff
@@ -1236,121 +1236,186 @@ __hs_delete_key_from_pos(
+            WT_ERR_MSG(session, WT_ERROR,
+              "the associated history store key %s was not found in the data store %s",
+              __wt_buf_set_printable(session, hs_key->data, hs_key->size, prev_hs_key),
+              session->dhandle->name);
```
Isn't this a panic?

sueloverso replied:
Fixed. The original had `WT_NOTFOUND`. Thanks. This should be a panic.

keithbostic commented on src/history/hs.c:
```diff
@@ -1236,121 +1236,186 @@ __hs_delete_key_from_pos(
+        ret = cursor->next(cursor);
+
+    /* If we positioned the cursor there is something to verify. */
+    if (ret != WT_NOTFOUND) {
```
I think this isn't safe, imagine `cursor.next` returned `WT_PANIC`.

sueloverso replied:
Fixed.

keithbostic commented on src/history/hs.c:
```diff
@@ -1236,121 +1236,186 @@ __hs_delete_key_from_pos(
+    cursor = session->hs_cursor;
+    ret = cursor->next(cursor);
+    stop = ret == WT_NOTFOUND ? true : false;
+    ret = 0;
```
This isn't safe, imagine `cursor.next` returned `WT_PANIC`.

sueloverso replied:
Fixed.

keithbostic commented on src/history/hs.c:
```diff
@@ -1236,121 +1236,186 @@ __hs_delete_key_from_pos(
+            WT_ERR_MSG(session, ret,
+              "Unable to find btree id %" PRIu32
+              " in the metadata file for the associated history store key %s",
+              btree_id, __wt_buf_set_printable(session, hs_key->data, hs_key->size, buf));
```
I think this is a panic?

sueloverso replied:
Fixed.

keithbostic commented on src/history/hs.c:
```diff
@@ -1236,121 +1236,186 @@ __hs_delete_key_from_pos(
+    uri_data = NULL;
+    is_owner = false; /* [-Wconditional-uninitialized] */
+
+    WT_ERR(__wt_scr_alloc(session, 0, &buf));
```
We don't really need `buf`, it's only used in an error path,  might as well just re-use one of the others, and not allocate a separate buffer.

sueloverso replied:
The only other buffer I have is `hs_key` which is in use in the error message.

keithbostic commented on src/meta/meta_table.c:
```diff
@@ -345,28 +345,20 @@ __wt_metadata_salvage(WT_SESSION_IMPL *session)
+     * Now verify the metadata. We know we're in wiredtiger_open and single threaded.
+     */
+    WT_RET(wt_session->verify(wt_session, WT_METAFILE_URI, NULL));
+    return (0);
```
I wouldn't bother having a separate function here given we know we're in `wiredtiger_open()`, but I wouldn't fight about it either.

sueloverso replied:
The only reason I put it there was to be co-located with `wt_metadata_salvage`. But that could also go away too.

sueloverso replied:
Fixed.

keithbostic commented on src/meta/meta_table.c:
```diff
@@ -345,28 +345,20 @@ __wt_metadata_salvage(WT_SESSION_IMPL *session)
+     * Now verify the metadata. We know we're in wiredtiger_open and single threaded.
+     */
+    WT_RET(wt_session->verify(wt_session, WT_METAFILE_URI, NULL));
+    return (0);
```
`return (wt_session->verify(wt_session, WT_METAFILE_URI, NULL));`

sueloverso replied:
This is no longer valid since I took the suggestion of not having a separate function.

keithbostic commented on src/utilities/util_main.c:
```diff
@@ -284,6 +291,11 @@ main(int argc, char *argv[])
+    /* If we only want to verify the metadata, that is done in wiredtiger_open. We're done. */
+    if (func == NULL && meta_verify)
+        goto done;
+
```
The comment above this line `/* Open the database and a session. */` probably needs to be split up since this line is in the middle (or removed entirely, it's not really adding a lot of value).

sueloverso replied:
Fixed.

keithbostic commented on src/utilities/util_verify.c:
```diff
@@ -30,17 +30,12 @@ util_verify(WT_SESSION *session, int argc, char *argv[])
+    dump_address = dump_blocks = dump_history = dump_layout = dump_pages = stable_timestamp = false;
     config = dump_offsets = uri = NULL;
-    while ((ch = __wt_getopt(progname, argc, argv, "ad:s")) != EOF)
+    while ((ch = __wt_getopt(progname, argc, argv, "d:s")) != EOF)
```
I think we need a corresponding fix in `src/docs/command-line.dox`.

sueloverso replied:
Good catch. Fixed.

sueloverso commented on dist/api_data.py:
```diff
@@ -1034,6 +1034,10 @@ def __ge__(self, other):
         ]),
+    Config('verify_metadata', 'false', r'''
+        open connection and verify any WiredTiger-owned table. This API
+        allows verification and detection of corruption in WiredTiger tables.''',
```
I agree with the sentiment and made this change. But I'd like to find some phrase other than `WiredTiger metadata` because that is something very specific in my head.

sueloverso commented on src/btree/bt_vrfy.c:
```diff
@@ -182,6 +182,16 @@ __wt_verify(WT_SESSION_IMPL *session, const char *cfg[])
+        skip_hs = true;
+    else
+        skip_hs = false;
+
```
Fixed.

sueloverso commented on src/btree/bt_vrfy.c:
```diff
@@ -268,6 +275,17 @@ __wt_verify(WT_SESSION_IMPL *session, const char *cfg[])
+                /* Open a history store cursor. */
+                WT_ERR(__wt_hs_cursor(session, &session_flags, &is_owner));
+                WT_TRET(__wt_history_store_verify_one(session));
+                WT_TRET(__wt_hs_cursor_close(session, session_flags, is_owner));
```
I was thinking we needed to do the later calls (exclusive and evicting the file with `WT_SYNC_DISCARD` and unloading the checkpoint) before calling WT_ERR. is it safe to skip all of those?

sueloverso replied:
I confirmed, I need to perform the other calls (at least the exclusive call) because if I don't and error out too soon, I trigger an assertion `[1587582440:623658][25469:0x7fbb7d7fa700], t, file:wt, WT_SESSION.verify: __wt_evict_file_exclusive_off, 944: v >= 0`. If I error out a few lines further down, then it panics correctly.

sueloverso replied:
I have added a comment explaining this at that location instead.

sueloverso commented on src/conn/conn_api.c:
```diff
@@ -2651,6 +2652,17 @@ wiredtiger_open(const char *home, WT_EVENT_HANDLER *event_handler, const char *c
+        ret = __wt_metadata_verify(verify_session);
+        wt_verify_session = &verify_session->iface;
+        WT_TRET(wt_verify_session->close(wt_verify_session, NULL));
+        WT_ERR(ret);
```
There's an assertion in hs.c that was failing when I used the default session and removed that assert. Now that the rest of the infrastructure is there, I'll try removing it again.

sueloverso replied:
I created WT-6048 to investigate this.

sueloverso commented on src/conn/conn_api.c:
```diff
@@ -2664,6 +2676,17 @@ wiredtiger_open(const char *home, WT_EVENT_HANDLER *event_handler, const char *c
+        wt_verify_session = &verify_session->iface;
+        WT_TRET(wt_verify_session->close(wt_verify_session, NULL));
+        WT_ERR(ret);
+    }
```
Good point. I will try to make that happen.

sueloverso replied:
I moved this and it seems to be working fine.

sueloverso commented on src/conn/conn_api.c:
```diff
@@ -2664,6 +2676,17 @@ wiredtiger_open(const char *home, WT_EVENT_HANDLER *event_handler, const char *c
 
     WT_ERR(__wt_metadata_cursor(session, NULL));
+    /*
+     * If the user wants to verify WiredTiger tables, verify the history store now that the metadata
```
Fixed.

sueloverso commented:
evergreen retry

sueloverso commented:
@keithbostic I did add the verify_metadata to `test/format` as your general comment suggested. I do that on `wiredtiger_open` anytime the verify flag is set. Mostly it won't have an effect since most format runs are not reopens. But anytime verify is used with reopen it should be tested.

keithbostic commented on src/history/hs.c:
```diff
@@ -1236,121 +1234,185 @@ __hs_delete_key_from_pos(
+        prev_hs_key = tmp;
     }
-    WT_ERR_NOTFOUND_OK(ret, false);
+    WT_ERR_NOTFOUND_OK(ret, true);
```
If we're not clearing `ret` and we're already at the `err:` lable, does this line actually do anything?

sueloverso replied:
You're correct, but I wanted to make it clear we should do this in case anyone was to add code after that line, but before the label.

keithbostic commented on src/history/hs.c:
```diff
@@ -1236,121 +1234,185 @@ __hs_delete_key_from_pos(
+        WT_ERR_NOTFOUND_OK(ret, false);
+    }
+    ret = 0;
+err:
```
This might be a little simpler as:
```
    ret = cursor->search_near(cursor, &exact);
    if (ret == 0 && exact < 0)  
        ret = cursor->next(cursor);
        
    /* If we positioned the cursor there is something to verify. */
    if (ret == 0) {
        __wt_btcur_init(session, &cbt);
        __wt_btcur_open(&cbt);
        ret = __verify_history_store_id(session, &cbt, btree_id);
        WT_TRET(__wt_btcur_close(&cbt, false));
    }
    return (ret == WT_NOTFOUND ? 0 : ret);
```
But I wouldn't argue about it.

sueloverso replied:
Fixed.

keithbostic commented on test/format/wts.c:
```diff
@@ -281,6 +281,9 @@ wts_open(const char *home, bool set_api, WT_CONNECTION **connp)
 
+    if (g.c_verify)
+        CONFIG_APPEND(p, ",verify_metadata=true");
+
```
I think this should just be on all the time, not conditional.

sueloverso replied:
I feel like if you don't want to verify, then you shouldn't. I think if one is doing a reopen and the HS file is huge, it could be a long time. I'm going to merge as-is. But that's as hard as I'm going to push back. If you feel strongly we can open a ticket and make this tiny change.
